Build a SHARED WebCore on WinCairo

donny-dont opened the pull request:
#### e912e7b00570fe6700dd49d366acc773616d0577
<pre>
Build a SHARED WebCore on WinCairo
<a href="https://bugs.webkit.org/show_bug.cgi?id=247944">https://bugs.webkit.org/show_bug.cgi?id=247944</a>

Reviewed by Fujii Hironori.

Make WebCore a `SHARED` library. Build PAL into it as well. Build
WebKitLegacy as a framework now that there isn&apos;t a conflict with where
WebCore is built.

Stub out `loadResourceIntoBuffer` which loads icons. It will be
addressed in a future bug.

* Source/WebCore/platform/graphics/win/ImageWin.cpp:
* Source/WebKit/WebProcess/win/WebProcessWin.cpp:
* Source/WebKitLegacy/CMakeLists.txt:
* Source/WebKitLegacy/PlatformWin.cmake:
* Source/WebKitLegacy/win/WebKitDLL.cpp:
* Source/cmake/OptionsWinCairo.cmake:

Canonical link: <a href="https://commits.webkit.org/257608@main">https://commits.webkit.org/257608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ffb00e943755d153daf37d39b09c676d696e675

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108800 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169037 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103412 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85924 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91903 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106722 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105173 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90481 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33910 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21828 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90098 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2470 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23345 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85956 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2389 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28900 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5232 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8546 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42820 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88823 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4257 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19868 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/win/ImageWin.cpp:
```diff
@@ -25,12 +25,20 @@
+// FIXME: https://bugs.webkit.org/show_bug.cgi?id=188175
+RefPtr<WebCore::FragmentedSharedBuffer> loadResourceIntoBuffer(const char*)
+{
+    return nullptr;
```
How many tests will fail for WinCairo WK1 by this change?

donny-dont replied:
Not sure but we will be turning off WK1 soon so not sure that we should be putting any effort on it.

fujii replied:
I think we should remove `loadResourceIntoBuffer` in Source/WebKit/WebProcess/win/WebProcessWin.cpp in this patch.
And, we should guard `loadResourceIntoBuffer` in Source/WebKitLegacy/win/WebKitDLL.cpp with `#if !PLATFORM(WIN_CAIRO)`.

fujii replied:
You added a link to <https://bugs.webkit.org/show_bug.cgi?id=188175>.
Do you plan to move icon resources into WebCore.dll?
Which directory will you use? In WebCore?
AppleWin is using static WebCore.
I think you can't move icon resources into WebCore directory. Right?

donny-dont replied:
Yea I plan on addressing this and hopefully finding a way to make it work for AppleWin and WinCairo its just not a blocker here cause its already broken in Modern WebKit on WinCairo.

fujii replied:
What is your motivation to make WebCore DLL? Only for the icon resource problem?

donny-dont replied:
No also you can't link a non-unified WebKit on Windows.

fujii reviewed (changes requested):
Needs a fix for https://github.com/WebKit/WebKit/pull/6519#discussion_r1042797583

donny-dont committed:
Build a SHARED WebCore on WinCairo
https://bugs.webkit.org/show_bug.cgi?id=247944

Reviewed by Fujii Hironori.

Make WebCore a `SHARED` library. Build PAL into it as well. Build
WebKitLegacy as a framework now that there isn't a conflict with where
WebCore is built.

Stub out `loadResourceIntoBuffer` which loads icons. It will be
addressed in a future bug.

* Source/WebCore/platform/graphics/win/ImageWin.cpp:
* Source/WebKit/WebProcess/win/WebProcessWin.cpp:
* Source/WebKitLegacy/CMakeLists.txt:
* Source/WebKitLegacy/PlatformWin.cmake:
* Source/WebKitLegacy/win/WebKitDLL.cpp:
* Source/cmake/OptionsWinCairo.cmake:

Canonical link: https://commits.webkit.org/257608@main
